File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/go/exercises/practice/two-bucket/two_bucket.go
package twobucket

import (
	"errors"
	"fmt"
)

type state struct {
	a, b  int
	steps int
}

func Solve(sizeBucketOne, sizeBucketTwo, goalAmount int, startBucket string) (string, int, int, error) {
	// Validate input parameters
	if sizeBucketOne <= 0 || sizeBucketTwo <= 0 || goalAmount <= 0 {
		return "", 0, 0, errors.New("invalid bucket sizes or goal")
	}
	if startBucket != "one" && startBucket != "two" {
		return "", 0, 0, errors.New("invalid start bucket")
	}
	if goalAmount > sizeBucketOne && goalAmount > sizeBucketTwo {
		return "", 0, 0, errors.New("goal exceeds both bucket capacities")
	}

	// Check GCD condition for impossibility
	if gcd(sizeBucketOne, sizeBucketTwo) != 0 && goalAmount%gcd(sizeBucketOne, sizeBucketTwo) != 0 {
		return "", 0, 0, errors.New("no solution exists")
	}

	// Check if initial fill matches goal
	if startBucket == "one" && sizeBucketOne == goalAmount {
		return "one", 1, 0, nil
	}
	if startBucket == "two" && sizeBucketTwo == goalAmount {
		return "two", 1, 0, nil
	}

	// Initialize BFS
	queue := []state{}
	visited := make(map[string]bool)
	if startBucket == "one" {
		queue = append(queue, state{sizeBucketOne, 0, 1})
		visited[fmt.Sprintf("%d,%d", sizeBucketOne, 0)] = true
	} else {
		queue = append(queue, state{0, sizeBucketTwo, 1})
		visited[fmt.Sprintf("%d,%d", 0, sizeBucketTwo)] = true
	}

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:]

		// Check if we've reached the goal
		if current.a == goalAmount {
			return "one", current.steps, current.b, nil
		}
		if current.b == goalAmount {
			return "two", current.steps, current.a, nil
		}

		// Generate all possible next states
		nextStates := []state{
			// Fill bucket 1
			{sizeBucketOne, current.b, current.steps + 1},
			// Fill bucket 2
			{current.a, sizeBucketTwo, current.steps + 1},
			// Empty bucket 1
			{0, current.b, current.steps + 1},
			// Empty bucket 2
			{current.a, 0, current.steps + 1},
			// Pour from 1 to 2
			{
				current.a - min(current.a, sizeBucketTwo-current.b),
				current.b + min(current.a, sizeBucketTwo-current.b),
				current.steps + 1,
			},
			// Pour from 2 to 1
			{
				current.a + min(current.b, sizeBucketOne-current.a),
				current.b - min(current.b, sizeBucketOne-current.a),
				current.steps + 1,
			},
		}

		for _, ns := range nextStates {
			// Validate state according to the rules
			if startBucket == "one" {
				if ns.a == 0 && ns.b == sizeBucketTwo {
					continue // Invalid state
				}
			} else {
				if ns.b == 0 && ns.a == sizeBucketOne {
					continue // Invalid state
				}
			}

			// Check if already visited
			key := fmt.Sprintf("%d,%d", ns.a, ns.b)
			if !visited[key] {
				visited[key] = true
				queue = append(queue, ns)
			}
		}
	}

	return "", 0, 0, errors.New("no solution possible")
}

func gcd(a, b int) int {
	for b != 0 {
		a, b = b, a%b
	}
	return a
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: src/2025-02-22-00-51-50--a-deepseek-r1-test-220-10-1--2-1-huoshan/python/exercises/practice/two-bucket/two_bucket.py
from collections import deque

def measure(bucket_one, bucket_two, goal, start_bucket):
    # Validate input parameters
    if bucket_one <= 0 or bucket_two <= 0:
        raise ValueError("Buckets must have positive capacity")
    if start_bucket not in ('one', 'two'):
        raise ValueError("start_bucket must be 'one' or 'two'")
    if goal < 0:
        raise ValueError("Goal must be non-negative")
    if goal > max(bucket_one, bucket_two):
        raise ValueError("Goal is impossible")

    # Determine capacities based on starting bucket
    start_cap = bucket_one if start_bucket == 'one' else bucket_two
    other_cap = bucket_two if start_bucket == 'one' else bucket_one

    # Check if goal exceeds both buckets' capacities
    if goal > max(start_cap, other_cap):
        raise ValueError("Goal exceeds capacity of both buckets")

    # Initialize BFS
    visited = set()
    queue = deque()
    
    # Create initial state based on starting bucket
    if start_bucket == 'one':
        initial_state = (bucket_one, 0)
    else:
        initial_state = (0, bucket_two)
    
    queue.append((initial_state[0], initial_state[1], 1))
    visited.add((initial_state[0], initial_state[1]))

    while queue:
        b1, b2, steps = queue.popleft()

        # Check if we've reached the goal
        if b1 == goal:
            return (steps, 'one', b2)
        if b2 == goal:
            return (steps, 'two', b1)

        # Generate all possible next states
        next_states = []

        # Action 1: Fill start bucket
        if start_bucket == 'one' and b1 < bucket_one:
            next_states.append((bucket_one, b2))
        elif start_bucket == 'two' and b2 < bucket_two:
            next_states.append((b1, bucket_two))

        # Action 2: Fill other bucket
        if start_bucket == 'one' and b2 < bucket_two:
            next_states.append((b1, bucket_two))
        elif start_bucket == 'two' and b1 < bucket_one:
            next_states.append((bucket_one, b2))

        # Action 3: Empty start bucket
        if start_bucket == 'one' and b1 > 0:
            next_states.append((0, b2))
        elif start_bucket == 'two' and b2 > 0:
            next_states.append((b1, 0))

        # Action 4: Empty other bucket
        if start_bucket == 'one' and b2 > 0:
            next_states.append((b1, 0))
        elif start_bucket == 'two' and b1 > 0:
            next_states.append((0, b2))

        # Action 5: Pour from start to other
        if start_bucket == 'one':
            pour = min(b1, bucket_two - b2)
            next_states.append((b1 - pour, b2 + pour))
        else:
            pour = min(b2, bucket_one - b1)
            next_states.append((b1 + pour, b2 - pour))

        # Action 6: Pour from other to start
        if start_bucket == 'one':
            pour = min(b2, bucket_one - b1)
            next_states.append((b1 + pour, b2 - pour))
        else:
            pour = min(b1, bucket_two - b2)
            next_states.append((b1 - pour, b2 + pour))

        # Process generated states
        for new_b1, new_b2 in next_states:
            # Check forbidden state rule
            if start_bucket == 'one':
                if new_b1 == 0 and new_b2 == bucket_two:
                    continue
            else:
                if new_b2 == 0 and new_b1 == bucket_one:
                    continue

            if (new_b1, new_b2) not in visited:
                visited.add((new_b1, new_b2))
                queue.append((new_b1, new_b2, steps + 1))

    # If we exhausted all possibilities
    raise ValueError("No solution exists")
